artemis-core-client/src/main/java/org/apache/activemq/artemis/core/client/impl/ServerLocatorImpl.java

```diff
@@ -1546,20 +1546,22 @@ public synchronized void connectorsChanged(List<DiscoveryEntry> newConnectors) {
       if (receivedTopology) {
          return;
       }
-      TransportConfiguration[] newInitialconnectors = (TransportConfiguration[]) Array.newInstance(TransportConfiguration.class, newConnectors.size());
 
-      int count = 0;
-      for (DiscoveryEntry entry : newConnectors) {
-         newInitialconnectors[count++] = entry.getConnector();
+      final List<TransportConfiguration> newInitialconnectors = new ArrayList<>(newConnectors.size());
 
+      for (DiscoveryEntry entry : newConnectors) {
          if (ha && topology.getMember(entry.getNodeID()) == null) {
             TopologyMemberImpl member = new TopologyMemberImpl(entry.getNodeID(), null, null, entry.getConnector(), null);
             // on this case we set it as zero as any update coming from server should be accepted
             topology.updateMember(0, entry.getNodeID(), member);
          }
+         // ignore its own transport connector
+         if (!entry.getConnector().equals(clusterTransportConfiguration)) {
```

Comment:
I think a core connection would get the topology sent to clients from the ClusterConnections's ServerLocator.

As long as your own connection is getting into the Topology for the clients, this is ok.

I know the test suite is passing, but did you check if the topology is being updated somewhere else? 


just wanted to make sure the test suite passed intentionally and not by an accident like other nodes would give you the update to the clients.

Reply:
> As long as your own connection is getting into the Topology for the clients, this is ok.

That's why the check `entry.getConnector().equals(clusterTransportConfiguration)` is placed after updating `topology`: the new `entry` should affect the topology. but not the `initialConnectors` list, because the latter would make it create a connection to itself ie is not a valid connector.

I'm just concerned that the we throw multiple times `ActiveMQClientMessageBundle.BUNDLE.noTCForSessionFactory()` that's a `ActiveMQException` and it will cause the `ConnectRunnable` to re-schedule itself on `retry interval`, see:
```java
   /**
    * used for making the initial connection in the cluster
    */
   private final class ConnectRunnable implements Runnable {

      private final ServerLocatorInternal serverLocator;

      private ConnectRunnable(ServerLocatorInternal serverLocator) {
         this.serverLocator = serverLocator;
      }

      @Override
      public void run() {
         try {
            if (started) {
               serverLocator.connect();
               if (serverLocator == replicationLocator) {
                  replicationClusterConnectedLatch.countDown();
               }
            }
         } catch (ActiveMQException e) {
            if (!started) {
               return;
            }
            if (logger.isDebugEnabled()) {

               logger.debug("retry on Cluster Controller " + System.identityHashCode(ClusterController.this) + " server = " + server);
            }
            server.getScheduledPool().schedule(this, serverLocator.getRetryInterval(), TimeUnit.MILLISECONDS);
         }
      }
   }
```
Throwing many exceptions isn't nice and I see that it will keep on doing it during the whole life-cycle of a live server *if* there is a single pair, because the backup won't have any acceptor opened to allow the live to connect to it...